datafusion/src/physical_plan/expressions/binary.rs

```diff
@@ -878,26 +934,90 @@ impl PhysicalExpr for BinaryExpr {
     }
 }
 
+/// The binary_array_op_scalar macro includes types that extend beyond the primitive,
+/// such as Utf8 strings.
+#[macro_export]
+macro_rules! binary_array_op_dyn_scalar {
+    ($LEFT:expr, $RIGHT:expr, $OP:ident) => {{
+        let is_numeric = DataType::is_numeric($LEFT.data_type());
+        let is_numeric_dict = match $LEFT.data_type() {
+            DataType::Dictionary(_, val_type) => DataType::is_numeric(val_type),
+            _ => false
+        };
+        let numeric_like = is_numeric | is_numeric_dict;
+
+        let is_string = ($LEFT.data_type() == &DataType::Utf8) | ($LEFT.data_type() == &DataType::LargeUtf8);
+        let is_string_dict = match $LEFT.data_type() {
+            DataType::Dictionary(_, val_type) => match **val_type {
+                DataType::Utf8 | DataType::LargeUtf8 => true,
+                _ => false
+            }
+        };
+        let string_like = is_string | is_string_dict;
+
+        let result: Result<Arc<dyn Array>> = if numeric_like {
+            compute_op_dyn_scalar!($LEFT, $RIGHT, $OP)
+        } else if string_like {
+            compute_utf8_op_dyn_scalar!($LEFT, $RIGHT, $OP)
+        } else {
+            let r: Result<Arc<dyn Array>> = match $LEFT.data_type() {
```

Comment:
@alamb FYI this is how i ended up implementing the pseudo code i previously mentioned.

```
if primitive_or_dict_vals_primitive {
    call_dyn_cmp_scalar(array, scalar, op)
} else if string_or_dict_vals_string {
    call_dyn_cmp_utf8_scalar(array, scalar, op)
} else {
    // Existing implementation for other types
}
```